### PR TITLE
feat(tidy3d): FXC-4043-add-support-of-custom-drc-args-in-drc-runner

### PR DESCRIPTION
**Pull request by marcorudolphflex**
Currently, there is no opportunity to pass custom args to the drc file. 

**Solution:**
We now can just provide a dict with some values which are then just appended with “-rd key=value” to the klayout drc command when using `DRCRunner.run(..., drc_args={key: value})`

<h2>Greptile Overview</h2>

Updated On: 2025-11-07 09:54:35 UTC

<h3>Greptile Summary</h3>

This PR adds support for passing custom arguments to KLayout DRC files through a new `drc_args` parameter in `DRCRunner.run()`. 

**Key Changes:**
- Added `drc_args: dict[str, str]` field to `DRCConfig` with comprehensive validation
- Arguments are forwarded to KLayout as command-line flags
- Validation ensures arguments are stringifiable and don't conflict with reserved keys
- Comprehensive test coverage including edge cases and error conditions

The implementation is well-designed with proper security considerations, thorough input validation, and excellent test coverage.

<h3>Confidence Score: 5/5</h3>

- This PR is safe to merge with minimal risk
- The implementation is well-designed with proper security considerations (list-based subprocess arguments prevent command injection), comprehensive input validation (type checking, reserved key protection, stringification), and excellent test coverage. All custom rules are followed correctly, and the change is backward compatible.
- No files require special attention

<h3>Important Files Changed</h3>

File Analysis

| Filename | Score | Overview |
|----------|-------|----------|
| tidy3d/plugins/klayout/drc/drc.py | 5/5 | Added `drc_args` field to `DRCConfig` with robust validation for stringifiable key/value pairs, reserved key checking, and proper command construction |
| tests/test_plugins/klayout/drc/test_drc.py | 5/5 | Comprehensive test coverage for the new `drc_args` feature including validation, stringification, reserved keys, and command construction |
| CHANGELOG.md | 5/5 | Added changelog entry documenting the new `drc_args` parameter in the klayout plugin |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant User
    participant DRCRunner
    participant DRCConfig
    participant run_drc_on_gds
    participant KLayout

    User->>DRCRunner: run(source, drc_args)
    DRCRunner->>DRCConfig: Create config with drc_args
    DRCConfig->>DRCConfig: _validate_drc_args_stringable()
    Note over DRCConfig: Convert keys/values to strings
    DRCConfig->>DRCConfig: _validate_drc_args_reserved()
    Note over DRCConfig: Check for reserved keys
    DRCConfig-->>DRCRunner: Validated config
    DRCRunner->>run_drc_on_gds: run_drc_on_gds(config)
    run_drc_on_gds->>run_drc_on_gds: Build command list
    Note over run_drc_on_gds: Add gdsfile parameter<br/>Add resultsfile parameter<br/>Add custom drc_args
    run_drc_on_gds->>KLayout: subprocess.run(cmd)
    KLayout-->>run_drc_on_gds: DRC results
    run_drc_on_gds->>DRCResults: load(resultsfile)
    DRCResults-->>run_drc_on_gds: Parsed results
    run_drc_on_gds-->>DRCRunner: DRCResults
    DRCRunner-->>User: DRCResults
```

### CODE REVIEW

File: tidy3d/plugins/klayout/drc/drc.py [hunk 1]
```diff
@@ -82,6 +88,56 @@ def _validate_drc_runset_format(cls, v: pd.FilePath) -> pd.FilePath:
...
+                value_str = str(value)
+            except Exception as exc:
+                raise ValidationError(
+                    f"DRC argument value {value!r} for key {key_str!r} of type "
+                    f"{type(value).__name__} cannot be converted to string: {exc}"
+                ) from exc
+
+            normalized[key_str] = value_str
+
+        return normalized
```
Comment (yaugenst-flex) [inline tidy3d/plugins/klayout/drc/drc.py]: Pretty sure all of this handled by pydantic via the field type?

Comment (marcorudolphflex) [reply tidy3d/plugins/klayout/drc/drc.py]: Yup, they are not coerced.

**Review by bzhangflex — APPROVED** [review]
